src/wordlist/wordlist.ts

```diff
@@ -0,0 +1,95 @@
+import {RBTree} from 'bintrees';
+
+import words from "@/wordlist/words.json";
+
+type WordTree = RBTree<string>;
+
+export const CODE_DELIMITER = '-';
+
+export class WordList {
+    private readonly _evenWords: string[];
+    private readonly _evenTree: WordTree;
+    private readonly _oddWords: string[];
+    private readonly _oddTree: WordTree;
+
+    constructor() {
+        // TODO: persist word list in a format that requires less processing.
+        this._evenWords = WordList._sideWords(0);
+        this._evenTree = new RBTree(WordList._less);
+        WordList._buildTree(this._evenTree, this._evenWords);
+
+        this._oddWords = WordList._sideWords(1);
+        this._oddTree = new RBTree(WordList._less);
+        WordList._buildTree(this._oddTree, this._oddWords);
+    }
+
+    public nextWordTree(previousWord: string): WordTree {
+        switch (previousWord) {
+            case '' || undefined:
+                return this._oddTree;
+        }
+
+        const foundOdd = typeof (this._oddTree.find(previousWord)) === 'string';
+        if (foundOdd) {
+            return this._evenTree;
+        }
+
+        const foundEven = typeof (this._evenTree.find(previousWord)) === 'string';
+        if (foundEven) {
+            return this._oddTree;
+        }
+
+        throw new Error('previous word not in even or odd word lists');
+    }
+
+    public get evenTree(): WordTree {
+        return this._evenTree;
+    }
+
+    public get oddTree(): WordTree {
+        return this._oddTree;
+    }
+
+    private static _buildTree(tree: WordTree, words: string[]) {
+        for (const word of words) {
+            tree.insert(word);
+        }
+    }
+
+    private static _less(a: any, b: any): 0 | -1 | 1 {
+        if (a === b) {
+            return 0;
+        }
+        return a < b ? -1 : 1;
+    }
+
+    private static _sideWords(side: 0 | 1): string[] {
+        const sideWords = [];
+        for (const index in words) {
+            sideWords.push((words as any)[index][side]);
+        }
+        return sideWords;
+    }
+}
+
+// NB: Odd code word comes first
+export class CodeCompleter {
+    private _wordList = new WordList();
+
+    public nearestNextWord(partialCode: string): string {
```

Comment:
I was expecting the type signature of this function to be `string -> [string]` since you get a bunch of words with the `partialCode` as a prefix and then we show those choices to the user and allow them to select one of those (or type more letters and narrow down the list to fewer possible words).

For example: typing `3-un` should generates these possible words - `3-un
3-undaunted-  3-underfoot-  3-unicorn-    3-unify-      3-universe-   3-unravel-`. At this point, there isn't any "nearest" word, only a bunch of words that start with the given prefix.

Reply:
The issue (#34) describes only showing a single match, which was the reason for only returning a single result.

It looks like the issue also describes showing a suggestion once there is only one match which is not the current behavior. I think we can determine if a word tree node is the only match by checking if there is a prefix match with the node's parent or its children. If not, that node is the only match.

Reply:
![PXL_20210930_115416154](https://user-images.githubusercontent.com/600733/135450718-8a8521c5-1a59-4583-846d-f5d1ccbf6896.jpg)

Reply:
(maybe it's sufficient just to check the children)

Reply:
Since we're finding the "lowerBound" I think we can simplify to only checking if the iterator's next node value has a matching prefix.